rfcs/20180918-functions-not-sessions-20.md

```diff
@@ -0,0 +1,1002 @@
+
+
+### A few things of note
+
+
+
+*   This change **only** applies to the TensorFlow **Python** frontend
+    *   [TensorFlow.js](https://js.tensorflow.org/) is already "eager by default".
+    *   [Switf4TensorFlow](https://github.com/tensorflow/swift) has [similar design goals](https://github.com/tensorflow/swift/blob/master/docs/DesignOverview.md#swift-for-tensorflow-design-overview), doing away with the define-then-run style of TensorFlow graphs.
+    *   Most other language bindings ([Java](https://www.tensorflow.org/api_docs/java/reference/org/tensorflow/package-summary), [C++](https://www.tensorflow.org/api_guides/cc/guide), [Go](https://godoc.org/github.com/tensorflow/tensorflow/tensorflow/go), others) are mostly targeting deployment of defined models in applications. While an imperative style might help simplify model development and training in these languages, doing so is explicitly out of scope for TensorFlow 2.0. The notion of graphs and sessions will remain in them, as well as in the stable [C API](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/c/c_api.h). In these APIs, the lifetime of program state (like variables) will continue to be tied to the lifetime of the `Session`.
+*   Users of **<code>Estimator</code>** will see no change
+    *   Canned Estimators are black boxes that create and train models. Enabling eager execution will have no effect on their usage. This is true today.
+    *   The <code>model_fn</code> of a regular (non-canned) <code>Estimator</code> will remain as a graph construction function.
+*   <strong>[SavedModel](https://www.tensorflow.org/guide/saved_model#save_and_restore_models)</strong> will continue to be the format encouraged for exporting trained models
+    *   Crudely speaking, a SavedModel encapsulates a Graph, a checkpoint of variable values, and some metadata like signature information (names of input and output tensors).
+    *   A path will be provided to easily export models in this format (e.g., via <code>tf.keras.Model.save()</code>). There may be instances where converting the Python code to a graph is not trivial, in which case, exporting to a SavedModel (and thus a Graph) will fail.
+
+
+## Alternatives Considered
+
+
+### Creating state inside a `defun`
+
+How state (`DT_RESOURCE` tensors)  created inside a `defun` should be handled is actively being debated. Options include:
+
+
+
+1.  "Lifting" state out as a static local function variable
+1.  Mimic the undecorated code - creating and destroying variables on each call.
+
```

Comment:
Did you consider just raising an exception? Why not just impose that functions decorated by `defun` should not create state? Can we list use cases that cannot be implemented in any other (simpler) ways?

Reply:
Great question. I gotta run now, but I'll respond to and let's continue discussion in the [thread above](https://github.com/tensorflow/community/pull/20#discussion_r219312959).

(Resolving this conversation to collapse with the thread above).